File: c1_algorithmic_toolbox/w2/fibonacci_last_digit.py
# Uses python3
import sys
import random


# Naive fibonacci last digit
def get_fibonacci_last_digit_naive(n):
    if n <= 1:
        return n

    previous = 0
    current = 1

    for _ in range(n - 1):
        previous, current = current, previous + current

    return current % 10


def get_fibonacci_last_digit_less_mem(n):
    if n <= 1:
        return n

    previous = 0
    current = 1

    for _ in range(n - 1):
        previous, current = current, (previous + current) % 10

    return current

if __name__ == '__main__':
    # input = sys.stdin.read()
    # n = int(input)
    # print(get_fibonacci_last_digit_naive(n))

    input = sys.stdin.read()
    n = int(input)
    print(get_fibonacci_last_digit_less_mem(n))

    # Stress test
    # while(True):
    #     n = random.randint(0, 10000)
    #     less_mem = get_fibonacci_last_digit_less_mem(n)
    #     print(less_mem)
    #     naive = get_fibonacci_last_digit_naive(n)
    #     print(naive)
    #     if (naive != less_mem):
    #         print('Wrong answer: naive: {} less_mem: {}\n'.
    #               format(naive, less_mem))
    #         break
    #     else:
    #         print('Ok!\n')

    # Overflow/Memory Test
    # n = int(1e7)
    # print(get_fibonacci_last_digit_less_mem(n))

File: c1_algorithmic_toolbox/w2/fibonacci_last_digit.cpp
#include <iostream>

int get_fibonacci_last_digit_less_mem(int n){
    if (n <=1) return n;
    else{
        long long previous = 0;
        long long current = 1;
        long long temp = 0;
        for(int i; i < n-1; ++i){
            temp = previous;
            previous = current;
            current = (temp+current) % 10;
        }
        return current;
    }
}

int main() {
    int n;
    std::cin >> n;
    int result;
    result = get_fibonacci_last_digit_less_mem(n);
    std::cout << result;
    return 0;
}